broker/service/spreadsheet_generation/spreadsheet_job_manager.py

```diff
@@ -0,0 +1,90 @@
+from dataclasses import dataclass
+from typing import Dict, BinaryIO, Optional
+import json
+from enum import Enum
+from concurrent.futures import ThreadPoolExecutor
+
+from broker.service.spreadsheet_generation.spreadsheet_generator import SpreadsheetGenerator, SpreadsheetSpec
+
+
+class JobStatus(Enum):
+    STARTED = "STARTED"
+    COMPLETE = "COMPLETE"
+    ERROR = "ERROR"
+
+
+@dataclass
+class JobSpec:
+    status: JobStatus
+    job_id: str
+    spreadsheet_path: str
+    filename: str
+
+    @staticmethod
+    def from_dict(data: Dict) -> 'JobSpec':
+        try:
+            return JobSpec(JobStatus[data["status"]], data["job_id"], data["spreadsheet_path"], data["filename"])
+        except:
+            raise
+
+    def to_dict(self) -> Dict:
+        return {
+            "status": self.status.value,
+            "job_id": self.job_id,
+            "spreadsheet_path": self.spreadsheet_path,
+            "filename": self.filename
+        }
+
+
+class SpreadsheetJobManager:
+    def __init__(self, spreadsheet_generator: SpreadsheetGenerator, output_dir_path: str, worker_pool: Optional[ThreadPoolExecutor]=None):
+        self.spreadsheet_generator = spreadsheet_generator
+        self.output_dir_path = output_dir_path
+        self.worker_pool = worker_pool if worker_pool is not None else ThreadPoolExecutor(5)
+
+    def create_job(self, spreadsheet_spec: SpreadsheetSpec, filename: str) -> JobSpec:
+        job_id = spreadsheet_spec.hashcode()
+        spreadsheet_output_path = f'{self.output_dir_path}/{job_id}.xlsx'
+        job_spec = JobSpec(JobStatus.STARTED, job_id, spreadsheet_output_path, filename)
+        job_spec_path = f'{self.output_dir_path}/{job_id}.json'
+
+        with open(job_spec_path, "w") as job_spec_file:
+            json.dump(job_spec.to_dict(), job_spec_file)
+
+        self.worker_pool.submit(lambda: self._do_create_spreadsheet_job(spreadsheet_spec, job_spec_path, spreadsheet_output_path))
+
+        return job_spec
+
+    def _do_create_spreadsheet_job(self, spreadsheet_spec: SpreadsheetSpec, job_spec_path: str, output_path: str):
+        job_result = self._maybe_create_spreadsheet(spreadsheet_spec, output_path)
+        job_spec = self.load_job_spec_from_path(job_spec_path)
+        completed_job_spec = JobSpec(job_result, job_spec.job_id, output_path, job_spec.filename)
+        self.write_job_spec(completed_job_spec, job_spec_path)
+
+    def _maybe_create_spreadsheet(self, spreadsheet_spec: SpreadsheetSpec, output_path: str) -> JobStatus:
+        try:
+            self.spreadsheet_generator.generate(spreadsheet_spec, output_path)
+            return JobStatus.COMPLETE
+        except Exception as e:
+            return JobStatus.ERROR
```

Comment:
It should be logged yep. Might be good idea to return an error the user as well, depending on what kind. There are certain kinds of errors that will be valuable e.g user specified an unknown type schema or module. Will add in logs and consider exceptions classes that would be valuable to the user

Reply:
it'll be tricky to add errors because there is a conversion happening from questionnaire data to actual list of modules. I think we should expect no errors on unknown schema/module because it should be implemented correctly from the UI. We need to identify other errors that can happen.

Reply:
Yep, we should not expect the users themselves to be able to do anything about the errors, but the error messages might be helpful for the maintainers of the UI, or for direct users of this API

Reply:
Ah yeah good point, i didn't realise there will be other users.